core/shared/src/main/scala/zio/internal/SingleThreadedRingBuffer.scala

```diff
@@ -51,7 +52,8 @@ private[zio] final class SingleThreadedRingBuffer[A <: AnyRef](capacity: Int) {
       result ::= array(i)
       i += 1
     }
-    result
+
+    result.filter(e => Option(e).isDefined)
```

Comment:
We can do this more efficiently if we just get each element in the `while` loop, check if it is `null`, and if not append it to the list. That way we only do a single traversal and don't allocate any additional data structures.